### PR TITLE
feat(cli-runtime): wrap meta.NoKindMatchErrors

### LINKED ISSUE #1515: cli-runtime/mapper: error not wrapped
### What happened?

When using the `resource.Builder` I want to ignore `meta.NoMatch*Error` errors like so:
```golang
resource.NewBuilder(..).
...
Stream(reader, "manifest").
...
Do.
IgnoreErrors(func(err error) bool {
    return meta.IsNoMatchError(err)
})
```
This does not work as the error is not wrapped in the `fmt.Errorf` [here](https://github.com/kubernetes/cli-runtime/blob/448da40f6f1654c1e6cca464fcad3b6fccaab0c7/pkg/resource/mapper.go#L69-L70) and I cannot match it to a `NoMatchError`

### What did you expect to happen?

I expect that the `meta.NoKindMatchError` error is wrapped and that `meta.IsNoMatchError(err)` returns **true**

### How can we reproduce it (as minimally and precisely as possible)?

Pass a manifest to `resource.Builder` which contains a resource for which the CRD has not been installed yet.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.2", GitCommit:"fc04e732bb3e7198d2fa44efa5457c7c6f8c0f5b", GitTreeState:"clean", BuildDate:"2023-02-22T13:32:21Z", GoVersion:"go1.20.1", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.5", GitCommit:"890a139214b4de1f01543d15003b5bda71aae9c7", GitTreeState:"clean", BuildDate:"2023-05-17T14:08:49Z", GoVersion:"go1.19.9", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
not applicable
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by IvoGoman**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Since go 1.13 errors can be wrapped. This PR adds wrapping for `meta.NoKindMatchError`, so that consumers of `resource.Builder` can choose to ignore this kind of error.
e.g.
```golang
resource.NewBuilder(..).
...
Stream(reader, "manifest").
...
Do.
IgnoreErrors(func(err error) bool {
    return meta.IsNoMatchError(err)
})
```
This removes the requirement to match on the error message itself, as reflected in the test.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1515

#### Does this PR introduce a user-facing change?
Not sure, if there is explicit matching on the error string then this might break such checks.
However, matching with `errors.Is(err, &meta.NoKindMatchError{})` should be the preferred way imo.

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go [hunk 1]
```diff
@@ -1050,7 +1051,7 @@ func TestRestMappingErrors(t *testing.T) {
 	// ensure that requesting a resource we _know_ not to exist results in an expected *meta.NoKindMatchError
 	err := b.Do().IntoSingleItemImplied(&singleItemImplied).Visit(test.Handle)
 	if err != nil {
-		if !strings.Contains(err.Error(), "server doesn't have a resource type \"foo\"") {
```
Comment (ardaguclu) [inline staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go:1053]: I think it is better to also have this error check to ensure that there is no behavioral change.

Comment (IvoGoman) [reply staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go:1053]: Makes sense. I have separated the checks to see at one glance if message or type is wrong

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go [hunk 2]
```diff
@@ -46,6 +46,7 @@ import (
 	"k8s.io/client-go/rest/fake"
 	restclientwatch "k8s.io/client-go/rest/watch"
 	"k8s.io/client-go/restmapper"
+
```
Comment (ardaguclu) [inline staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go:49]: nit: Redundant difference?

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go [hunk 3]
```diff
@@ -1051,7 +1052,10 @@ func TestRestMappingErrors(t *testing.T) {
 	err := b.Do().IntoSingleItemImplied(&singleItemImplied).Visit(test.Handle)
 	if err != nil {
 		if !strings.Contains(err.Error(), "server doesn't have a resource type \"foo\"") {
-			t.Fatalf("unexpected error: %v", err)
+			t.Fatalf("unexpected error message: %v", err)
```
Comment (ardaguclu) [inline staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go:1055]: Why did you change the current error message?

**Review by ardaguclu — COMMENTED** [review]
/ok-to-test

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go (hunk 2 above)
Comment (IvoGoman) [reply staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go:49]: Agree. This is done by gofmt. I would opt to leave this in as this will come up again the next time this file is touched.

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go (hunk 3 above)
Comment (IvoGoman) [reply staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go]: Reverted.

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go (hunk 2 above)
Comment (IvoGoman) [reply staging/src/k8s.io/cli-runtime/pkg/resource/builder_test.go]: Just removed it.

File: staging/src/k8s.io/cli-runtime/pkg/resource/builder.go [hunk 4]
```diff
@@ -803,7 +803,7 @@ func (b *Builder) mappingFor(resourceOrKindArg string) (*meta.RESTMapping, error
 		// if the error is _not_ a *meta.NoKindMatchError, then we had trouble doing discovery,
 		// so we should return the original error since it may help a user diagnose what is actually wrong
 		if meta.IsNoMatchError(err) {
-			return nil, fmt.Errorf("the server doesn't have a resource type %q", groupResource.Resource)
+			return nil, fmt.Errorf("the server doesn't have a resource type %q: %w", groupResource.Resource, err)
```
Comment (ardaguclu) [inline staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:806]: After investigating this carefully, I think we can't change this due to backwards compatibility reasons without breaking anywhere. Because it is one of the functions that widely used and probably some scripts in somewhere rely on this error message which is now we are changing.

Comment (IvoGoman) [reply staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:806]: There have been cases for similar errors where the handling was adopted: https://github.com/kubernetes/kubernetes/commit/dfcc141c91925c5cc7b04ed50bb9d51aa1b8b343

But I do agree changing error messages should be done with care. 
In this case I hoped it would be no issue, since the message is only appended to.

**Comment by ardaguclu** [comment]
Thanks
/triage accepted
/priority backlog
/approve
/lgtm

**Comment by IvoGoman** [comment]
Thank you for reviewing

**Review by brianpursley — APPROVED** [review]
/lgtm
/approve
